src/ZuzaluOracle.sol

```diff
@@ -1,65 +1,161 @@
+        }
+        if (_roots[1] != 0) {
+            _update($residentRoots, $residentsToDepth, _roots[1], _depths[1]);
+        }
+        if (_roots[2] != 0) {
+            _update($organizerRoots, $organizersToDepth, _roots[2], _depths[2]);
+        }
+        if (_roots[3] != 0) {
+            _update($participantRoots, $participantsToDepth, _roots[3], _depths[3]);
+        }
+        emit UpdateGroups(_roots, _depths);
     }
 
-    /// @notice
-    function _update(Commitment memory _commitment) internal {
-        _latestCommitment = _commitment;
-        commitments[_commitment.root] = _commitment.depth;
-        emit Update(_commitment.root, _commitment.depth);
+    /// @notice Updates the roots and depths of a particular group
+    function _update(
+        uint256[] storage _roots,
+        mapping(uint256 => uint256) storage _toDepth,
+        uint256 _root,
+        uint256 _depth
+    ) internal {
+        _roots.push(_root);
+        _toDepth[_root] = _depth;
     }
 
-    /// @notice
+    /*//////////////////////////////////////////////////////////////
+                                 VERIFY
+    //////////////////////////////////////////////////////////////*/
+
+    /// @notice Verifies a Semaphore proof for a particular signal and group. It returns true if the proof is valid, and false otherwise. It will check against
+    /// historic roots in case the latest root doesn't work. As the latest root may be updated between proof generation and verification, we want to offer a better
+    /// UX to the users by enabling by default to fall to check up to a limit of previous roots.
+    /// @notice It will check up to 2 roots in the past (latest + 2 roots).
+    /// @dev The backend updates the root every few hours, so at worse case the user will generate the proof and right away the backend will update the latest root. By checking the historic roots as well, we allow the user to still use the proof for a small extra cost in gas.
+    /// @param _nullifierHash The hash of the nullifier
+    /// @param _signal The signal to verify
+    /// @param _externalNullifier The external nullifier
+    /// @param _proof The proof to verify
+    /// @param _group The group to verify the proof for { None, Visitors, Residents, Organizers, Participants }
     function verify(
         uint256 _nullifierHash,
         uint256 _signal,
         uint256 _externalNullifier,
-        uint256[8] calldata _proof
-    ) external view returns (bool) {
-        uint256 root = _latestCommitment.root;
-        uint256 depth = _latestCommitment.depth;
+        uint256[8] calldata _proof,
+        Groups _group
+    ) external returns (bool) {
+        uint256 historicIndex = 1;
+        while (!_verifyHistoric(_nullifierHash, _signal, _externalNullifier, _proof, _group, historicIndex++)) {
+            if (historicIndex > 3) {
+                return false;
+            }
+        }
+        emit Verify(_signal, _group);
```

Comment:
Have you thought about the data that this event emits deeply? Just had the thought about also emitting the root that was used for the successful proof, at first thought it seems like a useful thing to emit

Reply:
No need to make this change, just thinking out loud